File: services/simulator/internal/app/audience.go
package app

import "media-planner/services/simulator/internal/domain"

type PublicSegment struct {
	ID      string `json:"segment_id"`
	Geo     string `json:"geo"`
	Gender  string `json:"gender"`
	AgeFrom int    `json:"age_from"`
	AgeTo   int    `json:"age_to_exclusive"`
}
type AudienceChannel struct {
	ChannelID domain.ChannelID `json:"channel_id"`
	Segments  []PublicSegment  `json:"segments"`
}
type AudienceCatalog struct {
	EngineVersion string            `json:"engine_version"`
	Digest        string            `json:"world_config_digest"`
	Channels      []AudienceChannel `json:"channels"`
}

func (r *Registry) AudienceCatalog() AudienceCatalog {
	result := AudienceCatalog{r.model.Model.EngineVersion, r.model.Digest, []AudienceChannel{}}
	for _, c := range r.model.Model.Channels {
		ch := AudienceChannel{c.ID, []PublicSegment{}}
		for _, s := range c.Segments {
			ch.Segments = append(ch.Segments, PublicSegment{s.ID, s.Geo, s.Gender, s.AgeFrom, s.AgeTo})
		}
		result.Channels = append(result.Channels, ch)
	}
	return result
}

File: services/frontend/src/domain/audience.ts
export interface ChannelAudience { segment_ids: string[] }
export type Audience = Record<string, ChannelAudience>
export interface AudienceSegment { segment_id: string; geo: string; gender: string; age_from: number; age_to_exclusive: number }
export interface AudienceCatalog { engine_version: string; world_config_digest: string; channels: { channel_id: string; segments: AudienceSegment[] }[] }
const idPattern = /^[a-z][a-z0-9_-]{0,63}$/
export function normalizeAudience(raw: unknown): Audience {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) throw new Error('invalid_audience')
  const entries = Object.entries(raw)
  if (!entries.length || entries.length > 20) throw new Error('invalid_audience')
  return Object.fromEntries(entries.sort(([a], [b]) => a.localeCompare(b)).map(([id, value]) => {
    if (!idPattern.test(id) || !value || typeof value !== 'object' || Array.isArray(value)) throw new Error('invalid_audience')
    const v = value as Record<string, unknown>
    if (Object.keys(v).some(key => !['segment_ids'].includes(key))) throw new Error('invalid_audience')
    if (!Array.isArray(v.segment_ids) || !v.segment_ids.length || v.segment_ids.length > 128 || v.segment_ids.some(s => typeof s !== 'string' || !idPattern.test(s)) || new Set(v.segment_ids).size !== v.segment_ids.length) throw new Error('invalid_audience')
    return [id, { segment_ids: [...v.segment_ids].sort() }]
  }))
}
export function audienceEqual(a: Audience | undefined, b: Audience | undefined): boolean {
  return JSON.stringify(a === undefined ? null : normalizeAudience(a)) === JSON.stringify(b === undefined ? null : normalizeAudience(b))
}
